@@ 
@@ Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
@@
@@ Licensed under the Apache License, Version 2.0 (the "License");
@@ you may not use this file except in compliance with the License.
@@ You may obtain a copy of the License at
@@ 
@@   http://www.apache.org/licenses/LICENSE-2.0
@@ 
@@ Unless required by applicable law or agreed to in writing, software
@@ distributed under the License is distributed on an "AS IS" BASIS,
@@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ See the License for the specific language governing permissions and
@@ limitations under the License.
@@ 
@@ For any questions about this software or licensing,
@@ please email opensource@seagate.com or cortx-questions@seagate.com.
@@

From 06e0d35ef026c69bb9fa8e788fa796009645cb81 Mon Sep 17 00:00:00 2001
Date: Wed, 3 Oct 2018 19:01:47 +0530
Subject: [PATCH] COSTOR-227: Changed config file and provisioned script with
 running only single instance.

---
 haproxy-statsd.conf        |  7 +++++++
 haproxy-statsd.conf.sample |  7 -------
 haproxy-statsd.py          | 27 +++++++++++++++++++++++++++
 3 files changed, 34 insertions(+), 7 deletions(-)
 create mode 100644 haproxy-statsd.conf
 delete mode 100644 haproxy-statsd.conf.sample

diff --git a/haproxy-statsd.conf b/haproxy-statsd.conf
new file mode 100644
index 0000000..95e1560
--- /dev/null
+++ b/haproxy-statsd.conf
@@ -0,0 +1,7 @@
+[haproxy-statsd]
+haproxy_url = http://localhost:8080/stats;csv
+haproxy_user =
+haproxy_password =
+statsd_host = localhost
+statsd_port = 8125
+statsd_namespace = haproxy.(HOSTNAME)
diff --git a/haproxy-statsd.conf.sample b/haproxy-statsd.conf.sample
deleted file mode 100644
index bd5db77..0000000
--- a/haproxy-statsd.conf.sample
+++ /dev/null
@@ -1,7 +0,0 @@
-[haproxy-statsd]
-haproxy_url = http://127.0.0.1:1936/;csv
-haproxy_user =
-haproxy_password =
-statsd_host = 127.0.0.1
-statsd_port = 8125
-statsd_namespace = haproxy.(HOSTNAME)
diff --git a/haproxy-statsd.py b/haproxy-statsd.py
index 945df02..7264d3f 100755
--- a/haproxy-statsd.py
+++ b/haproxy-statsd.py
@@ -28,6 +28,9 @@ import socket
 import argparse
 import ConfigParser
 import os
+import sys
+import fcntl
+import psutil
 
 import requests
 from requests.auth import HTTPBasicAuth
@@ -62,8 +65,27 @@ def report_to_statsd(stat_rows,
             stat_count += 1
     return stat_count
 
+def can_start():
+    lckfile = '/tmp/haproxy_statsd.lck'
+    try:
+        f = os.open(lckfile, os.O_RDWR|os.O_CREAT)
+        fcntl.flock(f, fcntl.LOCK_EX)
+        pid = os.read(f, 100)
+        if len(pid) > 0 and psutil.pid_exists(int(pid)):
+            return False
+
+        os.lseek(f, 0, 0)
+        os.write(f, '%s' %os.getpid())
+        os.fsync(f)
+        fcntl.flock(f, fcntl.LOCK_UN)
+        os.close(f)
+        return True
+
+    except Exception as e:
+        return False
 
 if __name__ == '__main__':
+
     parser = argparse.ArgumentParser(
         description='Report haproxy stats to statsd')
     parser.add_argument('-c', '--config',
@@ -87,6 +109,11 @@ if __name__ == '__main__':
     config.add_section('haproxy-statsd')
     config.read(args.config)
 
+    # Check if script is already running
+    if not can_start():
+        # Script is already running therefore exit
+        sys.exit(1)
+
     # Generate statsd namespace
     namespace = config.get('haproxy-statsd', 'statsd_namespace')
     if '(HOSTNAME)' in namespace:
-- 
1.8.3.1

